lib/ABIAnalyses/DeadReturnValuesOfFunctionCall.cpp

```diff
@@ -0,0 +1,52 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/GlobalVariable.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Support/Casting.h"
+
+#include "revng/ABIAnalyses/Analyses.h"
+#include "revng/MFP/MFP.h"
+#include "revng/Model/Binary.h"
+#include "revng/Support/revng.h"
+
+namespace ABIAnalyses::DeadReturnValuesOfFunctionCall {
+using namespace llvm;
+using namespace ABIAnalyses;
+
+DenseMap<const GlobalVariable *, State>
+analyze(const BasicBlock *CallSiteBlock, const GeneratedCodeBasicInfo &GCBI) {
+  using MFI = MFI<true>;
+
+  MFI Instance{ { getPreCallHook(CallSiteBlock), GCBI } };
+  MFI::LatticeElement InitialValue{};
+  MFI::LatticeElement ExtremalValue{};
+  auto *Start = CallSiteBlock->getUniqueSuccessor();
+  auto
+    Results = MFP::getMaximalFixedPoint<MFI, MFI::GT, MFI::LGT>(Instance,
+                                                                Start,
+                                                                InitialValue,
+                                                                ExtremalValue,
+                                                                { Start },
+                                                                { Start });
+
+  DenseMap<const GlobalVariable *, State> RegNoOrDead{};
+
+  for (auto &[BB, Result] : Results) {
+    for (auto &[GV, RegState] : Result.OutValue) {
+      if (RegState == CoreLattice::NoOrDead) {
+        RegNoOrDead[GV] = State::NoOrDead;
+      }
+    }
+  }
+  return RegNoOrDead;
```

Comment:
Use `Binary.h` enum.
I was under the impression that this branch has been rebased.

Reply:
yes, so the intermediate results use `CoreLattice::LatticeElement` enums that are generated from the `.dot` file spec

The result is `State` which is `model::RegisterState::Values` from `Binary.h`

The reason is that the lattice contains states that are not in the model and may not make sense to add to the model (i.e. `Bottom` which is in some analyses)

Reply:
Can't you use `Invalid` as `Bottom`?

Reply:
makes sense, then, if we are going that way, I propose to use `Invalid` instead of `Bottom` and to add `Unknown` as a register state, would that work?

Reply:
I forgot about `Unknown`. Mh, I don't want to expose an implementation detail to the Model, which is our long term disk format.
Keep things as they are.